examples/gateway-reshard.rs

```diff
@@ -0,0 +1,122 @@
+        .collect::<Vec<_>>()
+        .await;
+
+    loop {
+        // Run `gateway_runner` and `reshard` concurrently until one finishes.
+        tokio::select! {
+            // Gateway_runner only finises on errors, so break the outer loop.
+            _ = gateway_runner(shards, Arc::clone(&client)) => break,
+            // Resharding done! Time to run `gateway_runner` with the new list
+            // of shards.
+            Ok(Some(new_shards)) = reshard(&client, config_callback) => {
+                    shards = new_shards;
+            },
+        }
+    }
+
+    Ok(())
+}
+
+async fn gateway_runner(mut shards: Vec<Shard>, client: Arc<Client>) {
+    let mut stream = ShardEventStream::new(shards.iter_mut());
+
+    loop {
+        let event = match stream.next().await {
+            Some((_, Ok(event))) => event,
+            Some((_, Err(source))) => {
+                tracing::warn!(?source, "error receiving event");
+
+                if source.is_fatal() {
+                    break;
+                }
+
+                continue;
+            }
+            None => break,
+        };
+
+        tokio::spawn(event_handler(Arc::clone(&client), event));
+    }
+}
+
+async fn event_handler(client: Arc<Client>, event: Event) -> anyhow::Result<()> {
+    match event {
+        Event::MessageCreate(message) if message.content == "!ping" => {
+            client
+                .create_message(message.channel_id)
+                .content("Pong!")?
+                .exec()
+                .await?;
+        }
+        _ => {}
+    }
+
+    Ok(())
+}
+
+/// Tries to create a list of the recommended amount of shards.
+///
+/// Returns when all shards are identified.
+async fn reshard(
+    client: &Client,
+    config_callback: impl Fn(ShardId) -> Config,
+) -> anyhow::Result<Option<Vec<Shard>>> {
+    const RESHARD_DURATION: Duration = Duration::from_secs(60 * 60 * 8);
+
+    // reshard every eight hours
+    time::sleep(RESHARD_DURATION).await;
+
+    let mut shards = stream::start_recommended(&client, &config_callback)
+        .await?
+        .filter_map(|shard_result| async move { shard_result.ok() })
+        .collect::<Vec<_>>()
+        .await;
+
+    let mut identified = vec![false; shards.len()];
+    // don't deserialize the events (with `ShardEventStream`) as they're
+    // duplicates of the already running shards
+    let mut stream = ShardMessageStream::new(shards.iter_mut());
+
+    // drive the set of shards until they are all identified
+    while !identified.iter().all(|&shard| shard) {
+        match stream.next().await {
+            Some((_, Err(source))) => {
+                tracing::warn!(?source, "error receiving event");
+
+                if source.is_fatal() {
+                    return Ok(None);
+                }
+
+                continue;
+            }
+            Some((shard, _)) => {
+                identified[shard.id().number() as usize] = shard.session().is_some();
```

Comment:
`shard.session().is_some()` is kinda incorrect, but it's the best proxy for checking the shard's identification status (more variants on `ConnectionStatus` would be ideal)